x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java

```diff
@@ -0,0 +1,160 @@
+        Map<String, OsStats> osStatsPerNode,
+        ActionListener<MlAutoscalingStats> listener
+    ) {
+        Set<String> nodesWithRunningJobs = new HashSet<>();
+        long memoryBytesSum = osStatsPerNode.values()
+            .stream()
+            .map(s -> s.getMem().getAdjustedTotal().getBytes())
+            .collect(Collectors.summingLong(Long::longValue));
+
+        long modelMemoryBytesSum = 0;
+        long extraSingleNodeModelMemoryInBytes = 0;
+        int extraSingleNodeProcessors = 0;
+        int minNodes = 0;
+        int extraProcessors = 0;
+
+        final MlAutoscalingContext autoscalingContext = new MlAutoscalingContext(clusterState);
+
+        for (var task : autoscalingContext.anomalyDetectionTasks) {
+            String jobId = ((OpenJobAction.JobParams) task.getParams()).getJobId();
+            if (AWAITING_LAZY_ASSIGNMENT.equals(task.getAssignment())) {
+                extraSingleNodeModelMemoryInBytes = Math.max(
+                    extraSingleNodeModelMemoryInBytes,
+                    mlMemoryTracker.getAnomalyDetectorJobMemoryRequirement(jobId)
+                );
+                extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, 1);
+                ++extraProcessors;
+            } else {
+                modelMemoryBytesSum += mlMemoryTracker.getAnomalyDetectorJobMemoryRequirement(jobId);
+                nodesWithRunningJobs.add(task.getExecutorNode());
+                minNodes = 1;
+            }
+        }
+        for (var modelAssignment : autoscalingContext.modelAssignments.entrySet()) {
+            if (AssignmentState.STARTING.equals(modelAssignment.getValue().getAssignmentState())
+                && modelAssignment.getValue().getNodeRoutingTable().isEmpty()) {
+                extraSingleNodeModelMemoryInBytes = Math.max(
+                    extraSingleNodeModelMemoryInBytes,
+                    modelAssignment.getValue().getTaskParams().estimateMemoryUsageBytes()
+                );
+                final int unassignedProcessors = modelAssignment.getValue().getTaskParams().getNumberOfAllocations() * modelAssignment
+                    .getValue()
+                    .getTaskParams()
+                    .getThreadsPerAllocation();
+
+                extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, unassignedProcessors);
+
+                extraProcessors += unassignedProcessors;
+            } else {
+                modelMemoryBytesSum += modelAssignment.getValue().getTaskParams().estimateMemoryUsageBytes();
+
+                final int assignedProcessors = modelAssignment.getValue().getTaskParams().getNumberOfAllocations() * modelAssignment
+                    .getValue()
+                    .getTaskParams()
+                    .getThreadsPerAllocation();
+
+                // min(3, max(number of allocations over all deployed models), as this is always 3
+                minNodes = Math.min(3, Math.max(minNodes, assignedProcessors));
```

Comment:
This formula should use number of allocations rather than assigned processors. For example, 1 allocation with 3 threads per allocation should only require 1 node and not 3.

Also, should `minNodes` be set in the same way even if the only models are awaiting allocation?